### PR TITLE
Use spatial ref information

### PR DESCRIPTION
**Pull request by jdroenner**
This PR
- adds checks in wms and wfs services: a) fail if request has no crs, b) fail if request and workflow crs are not equal
- renames wgs84 to epsg_4326
- enables crs reading in the gdal_source

### CODE REVIEW

File: services/src/error.rs [hunk 1]
```diff
@@ -66,6 +67,13 @@ pub enum Error {
 
     InvalidNamespace,
 
+    InvalidSpatialReference,
+    #[snafu(display("SpatialReferenceMissmatch {} != {}", a, b))]
+    SpatialReferenceMissmatch {
+        a: SpatialReferenceOption,
+        b: SpatialReferenceOption,
+    },
+
```
Comment (ChristianBeilschmidt) [inline services/src/error.rs:76]: You could maybe also rename the error.

And I would say `expected: {}, found: {}` or does the order not matter for the error and the user?

File: services/src/handlers/wfs.rs [hunk 2]
```diff
@@ -189,6 +189,19 @@ async fn get_feature<C: Context>(
         .initialize(&execution_context)
         .context(error::Operator)?;
 
+    // handle request and workflow crs matching
+    let spatial_ref = initialized.result_descriptor().spatial_reference();
+    // TODO: use a default spatial reference if it is not set?
+    snafu::ensure!(request.srs_name.is_some(), error::InvalidSpatialReference);
```
Comment (ChristianBeilschmidt) [inline services/src/handlers/wfs.rs:195]: 4326 by default?

File: services/src/handlers/wfs.rs [hunk 3]
```diff
@@ -189,6 +189,19 @@ async fn get_feature<C: Context>(
...
+    // TODO: inject projection Operator?
```
Comment (ChristianBeilschmidt) [inline services/src/handlers/wfs.rs:196]:
```suggestion
    // TODO: inject projection Operator
```
:+1:

File: services/src/handlers/wms.rs [hunk 4]
```diff
@@ -169,6 +170,19 @@ async fn get_map<C: Context>(
         .initialize(&execution_context)
         .context(error::Operator)?;
 
+    // handle request and workflow crs matching
+    let spatial_ref = initialized.result_descriptor().spatial_reference();
+    // TODO: use a default spatial reference if it is not set?
+    snafu::ensure!(request.crs.is_some(), error::InvalidSpatialReference);
+    // TODO: inject projection Operator?
```
Comment (ChristianBeilschmidt) [inline services/src/handlers/wms.rs:177]:
```suggestion
    // TODO: inject projection Operator
```
:+1:

File: services/src/handlers/wms.rs [hunk 5]
```diff
@@ -169,6 +170,19 @@ async fn get_map<C: Context>(
         .initialize(&execution_context)
         .context(error::Operator)?;
 
+    // handle request and workflow crs matching
+    let spatial_ref = initialized.result_descriptor().spatial_reference();
+    // TODO: use a default spatial reference if it is not set?
```
Comment (ChristianBeilschmidt) [inline services/src/handlers/wms.rs:175]: also here, default to 4326

File: services/src/ogc/wms/request.rs [hunk 6]
```diff
@@ -48,7 +51,7 @@ pub struct GetMap {
     #[serde(alias = "LAYERS")]
     pub layers: String,
     #[serde(alias = "CRS")]
-    pub crs: String, // TODO: parse CRS
+    pub crs: Option<SpatialReference>,
```
Comment (ChristianBeilschmidt) [inline services/src/ogc/wms/request.rs:54]: Doesn't `SpatialReferenceOption` work?
We could use no sref if it works instead of a default?

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: services/src/handlers/wms.rs (hunk 5 above)
Comment (jdroenner) [reply services/src/handlers/wms.rs:175]: The GeoServer docs state crs is a required parameter. So it makes no sense to allow it to be empty. 
https://docs.geoserver.org/stable/en/user/services/wms/reference.html

File: services/src/handlers/wfs.rs (hunk 2 above)
Comment (jdroenner) [reply services/src/handlers/wfs.rs:195]: i really don't know. i guess the default is to use the projection of the dataset or in our case the workflow

File: services/src/ogc/wms/request.rs (hunk 6 above)
Comment (jdroenner) [reply services/src/ogc/wms/request.rs:54]: yeah well it was like this in wfs so i just adapted wms. I can change it if you prefer it.

File: services/src/error.rs (hunk 1 above)
Comment (jdroenner) [reply services/src/error.rs:76]: done

File: services/src/handlers/wms.rs (hunk 4 above)
Comment (jdroenner) [reply services/src/handlers/wms.rs:177]: new task

File: services/src/handlers/wfs.rs (hunk 3 above)
Comment (jdroenner) [reply services/src/handlers/wfs.rs:196]: new task

File: services/src/error.rs (hunk 1 above)
Comment (jdroenner) [reply services/src/error.rs]: 29a9e1e

File: services/src/handlers/wfs.rs (hunk 3 above)
Comment (ChristianBeilschmidt) [reply services/src/handlers/wfs.rs]: Can you add it as a task somewhere? There are multiple tasks regarding projections (story?). Then, I can close it.

Comment (ChristianBeilschmidt) [reply services/src/handlers/wfs.rs]: https://trello.com/c/ue8BMo64/255-allow-users-to-re-project-data

**Review by ChristianBeilschmidt — APPROVED** [review]

**Pull request merged**